v2/internal/reconcilers/arm/arm_client_cache.go

```diff
@@ -82,37 +85,96 @@ func (c *armClientCache) lookup(key string) (*armClient, bool) {
+	if err != nil {
+		return nil, err
 	}
 
 	if secret != nil {
-		nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
-		client, ok := c.lookup(nsName.String())
-		// if client does not already exist, or secret data is changed, we need to get new token.
-		if !ok || !matchSecretData(client.secretData, secret.Data) {
-			credential, subscriptionID, err := c.newCredentialFromSecret(secret, nsName)
-			if err != nil {
-				return nil, "", err
-			}
-
-			newClient, err := genericarmclient.NewGenericClientFromHTTPClient(c.cloudConfig, credential, c.httpClient, subscriptionID, metrics.NewARMClientMetrics())
-			if err != nil {
-				return nil, "", err
-			}
-
-			armClient := newARMClient(newClient, secret.Data, nsName)
-			c.register(armClient)
-			return armClient.GenericClient(), armClient.CredentialFrom(), nil
-		} else {
-			return client.GenericClient(), client.CredentialFrom(), nil
+		armClient, err := c.getARMClientFromSecret(secret)
+		if err != nil {
+			return nil, err
 		}
+		return armClient, nil
 	}
+	return nil, nil
+}
 
-	// If not found, default would be global
-	return c.globalClient.GenericClient(), c.globalClient.CredentialFrom(), nil
+func (c *armClientCache) checkPerResourceSecret(ctx context.Context, obj genruntime.ARMMetaObject) (*armClient, error) {
+	credentialFrom, ok := obj.GetAnnotations()[perResourceSecretAnnotation]
+	// annotation exists, use per Resource Group secret credential
+	if ok {
+		secretNamespacedName := getSecretNameFromAnnotation(credentialFrom, obj.GetNamespace())
+
+		secret, err := c.getSecret(ctx, secretNamespacedName.Namespace, secretNamespacedName.Name)
+		if err != nil {
+			return nil, err
+		}
+		// secret not found
+		if secret == nil {
+			return nil, core.NewSecretNotFoundError(secretNamespacedName, errors.Errorf("Credential Secret %q not found for '%s/%s'", secretNamespacedName.String(), obj.GetNamespace(), obj.GetName()))
+		}
+
+		armClient, err := c.getARMClientFromSecret(secret)
+		if err != nil {
+			return nil, err
+		}
+		return armClient, nil
+	}
+	return nil, nil
+}
+
+func getSecretNameFromAnnotation(credentialFrom string, resourceNamespace string) types.NamespacedName {
+	if strings.Contains(credentialFrom, namespacedNameSeparator) {
+		slice := strings.Split(credentialFrom, namespacedNameSeparator)
+		return types.NamespacedName{Namespace: slice[0], Name: slice[1]}
+	} else {
+		// If namespace is not specified, look into the resource namespace
+		return types.NamespacedName{Namespace: resourceNamespace, Name: credentialFrom}
+	}
+}
+
+func (c *armClientCache) getARMClientFromSecret(secret *v1.Secret) (*armClient, error) {
+	nsName := types.NamespacedName{Namespace: secret.Namespace, Name: secret.Name}
+	client, ok := c.lookup(nsName.String())
+	// if client does not already exist, or secret data is changed, we need to get new token.
+	if !ok || !matchSecretData(client.secretData, secret.Data) {
```

Comment:
minor: You could also remove nesting here if you wanted via by swapping the order of these:

```
	client, ok := c.lookup(nsName.String())
	if ok && matchSecretData(client.secretData, secret.Data) {
		return client, nil
	}

	// Other stuff you had in the if

```